### ISSUE TITLE
App crashes when backup data is restored from Google Drive

### ISSUE THREAD

**Issue by lchnnl** [forcedotcom/SalesforceMobileSDK-Android#2074]
We have noticed that during the installation of a new update, the app crashes when Android Automatic Backup functionality is turned **on**. The app crashes when the restore functionality has finished running. The app will immediately crash upon opening. This seems to be a similar issue to #2018 

We are turning off automatic backup with android:allowBackup="false" application attribute in the AndroidManifest.xml. That seems to do the trick for the short term. The downside is that login credentials are all lost, pincode is also lost. For apps with frequent updates that is not ideal, because it would require the user to login again, and recreate the pincode everything there is a new update.

Please fill out the following details:

**1. Version of Mobile SDK Used:** v7.2.0
**2. Issue found in Native App or Hybrid App:** Hybrid
**3. OS Version:** Android 9.0
**4. Device:** Simulator with Google Play.
**5. Steps to reproduce:**

Pre requisite: Logged into google account on your device. Turned on automatic backup according to the documentation: [https://developer.android.com/guide/topics/data/testingbackup](https://developer.android.com/guide/topics/data/testingbackup)
1. Open the app.
2. Manually trigger the backup functionality
3. Close the app again
4. Rebuild and reinstall the app
5. Manually trigger the restore functionality as described according to the documentation
6. App crashes when reopening.

**6. Actual behavior:**
App crashes when reopening

**7. Expected Behavior:**
App does not crash when reopening

**8. Error Log:**
```
D/SalesforceHybridSDKManager: Setting up global store using config found in www/globalstore.json
E/Encryptor: Error during asymmetric decryption
    javax.crypto.IllegalBlockSizeException
        at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:519)
        at javax.crypto.Cipher.doFinal(Cipher.java:2055)
        at com.salesforce.androidsdk.analytics.security.Encryptor.decryptWithPrivateKey(Encryptor.java:365)
        at com.salesforce.androidsdk.analytics.security.Encryptor.decryptWithRSABytes(Encryptor.java:318)
        at com.salesforce.androidsdk.analytics.security.Encryptor.decryptWithRSA(Encryptor.java:299)
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.generateUniqueId(SalesforceKeyGenerator.java:222)
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getUniqueId(SalesforceKeyGenerator.java:90)
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getUniqueId(SalesforceKeyGenerator.java:79)
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.generateEncryptionKey(SalesforceKeyGenerator.java:203)
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getEncryptionKey(SalesforceKeyGenerator.java:102)
        at com.salesforce.androidsdk.app.SalesforceSDKManager.getEncryptionKey(SalesforceSDKManager.java:1129)
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCurrentAccount(UserAccountManager.java:223)
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCurrentUser(UserAccountManager.java:172)
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCachedCurrentUser(UserAccountManager.java:186)
        at com.salesforce.androidsdk.analytics.EventBuilderHelper.createAndStore(EventBuilderHelper.java:120)
        at com.salesforce.androidsdk.analytics.EventBuilderHelper.access$000(EventBuilderHelper.java:45)
        at com.salesforce.androidsdk.analytics.EventBuilderHelper$1.run(EventBuilderHelper.java:92)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: android.security.KeyStoreException: Unknown error
        at android.security.KeyStore.getKeyStoreException(KeyStore.java:839)
        at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:224)
        at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:506)
        at javax.crypto.Cipher.doFinal(Cipher.java:2055) 
        at com.salesforce.androidsdk.analytics.security.Encryptor.decryptWithPrivateKey(Encryptor.java:365) 
        at com.salesforce.androidsdk.analytics.security.Encryptor.decryptWithRSABytes(Encryptor.java:318) 
        at com.salesforce.androidsdk.analytics.security.Encryptor.decryptWithRSA(Encryptor.java:299) 
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.generateUniqueId(SalesforceKeyGenerator.java:222) 
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getUniqueId(SalesforceKeyGenerator.java:90) 
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getUniqueId(SalesforceKeyGenerator.java:79) 
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.generateEncryptionKey(SalesforceKeyGenerator.java:203) 
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getEncryptionKey(SalesforceKeyGenerator.java:102) 
        at com.salesforce.androidsdk.app.SalesforceSDKManager.getEncryptionKey(SalesforceSDKManager.java:1129) 
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCurrentAccount(UserAccountManager.java:223) 
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCurrentUser(UserAccountManager.java:172) 
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCachedCurrentUser(UserAccountManager.java:186) 
        at com.salesforce.androidsdk.analytics.EventBuilderHelper.createAndStore(EventBuilderHelper.java:120) 
        at com.salesforce.androidsdk.analytics.EventBuilderHelper.access$000(EventBuilderHelper.java:45) 
        at com.salesforce.androidsdk.analytics.EventBuilderHelper$1.run(EventBuilderHelper.java:92) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
E/SalesforceKeyGenerator: Exception thrown while getting encryption key
    java.lang.NullPointerException: Attempt to invoke virtual method 'byte[] java.lang.String.getBytes(java.nio.charset.Charset)' on a null object reference
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.generateEncryptionKey(SalesforceKeyGenerator.java:204)
        at com.salesforce.androidsdk.security.SalesforceKeyGenerator.getEncryptionKey(SalesforceKeyGenerator.java:102)
        at com.salesforce.androidsdk.app.SalesforceSDKManager.getEncryptionKey(SalesforceSDKManager.java:1129)
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCurrentAccount(UserAccountManager.java:223)
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCurrentUser(UserAccountManager.java:172)
        at com.salesforce.androidsdk.accounts.UserAccountManager.getCachedCurrentUser(UserAccountManager.java:186)
        at com.salesforce.androidsdk.analytics.EventBuilderHelper.createAndStore(EventBuilderHelper.java:120)
        at com.salesforce.androidsdk.analytics.EventBuilderHelper.access$000(EventBuilderHelper.java:45)
        at com.salesforce.androidsdk.analytics.EventBuilderHelper$1.run(EventBuilderHelper.java:92)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
I/Process: Sending signal. PID: 26052 SIG: 9
```

**Comment by bhariharan** [comment]
@lchnnl Yes, we don't support backup to Google Drive in any of our applications at the moment for security reasons. The encryption key is created on the fly when the app is installed, so restoring data wouldn't work since the old encryption key isn't available anymore.

**Comment by lchnnl** [comment]
Hi @bhariharan 
Thank you for your prompt reply and explanation. If backup are not supported by the SDK, then it should be turned off by default. This has not been turned off in the cordova plugin:

https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin/blob/c1570b4cb1796e75e6befb5022574a84525b74c4/plugin.xml#L157

I would at least expect an attribute `android:allowBackup="false"`

**Comment by wmathurin** [comment]
We will make the change.
We have the correct configuration on the iOS side: https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin/blob/c1570b4cb1796e75e6befb5022574a84525b74c4/plugin.xml#L50

**Comment by wmathurin** [comment]
Fix coming to our next release (8.2): https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin/pull/515
